### ISSUE TITLE
Lint/SafeNavigationChain and `try`

### ISSUE THREAD

**Issue by MOZGIII** [bbatsov/rubocop#3915]
Cop `Lint/SafeNavigationChain` does not allow `try` and `try!`.

--------

## Expected behavior

`try` and `try!` should be allowed after `&.` in Rails.

## Actual behavior

Cop does not allow `try` and `try!` after `&.` in Rails.

## Steps to reproduce the problem

```
$ echo 'a&.b.try(:c)' | rubocop -s demo
Inspecting 1 file
W

Offenses:

demo:1:5: W: Do not chain ordinary method call after safe navigation operator.
a&.b.try(:c)
    ^

1 file inspected, 1 offense detected
```

## RuboCop version
```
$ rubocop -V
0.47.0 (using Parser 2.3.3.1, running on ruby 2.3.1 x86_64-linux)
```

**Comment by bbatsov** [comment]
I see no reason why someone should write code like this. Some rationale would be useful...

**Comment by MOZGIII** [comment]
Here: `property&.params.try(:purpose)&.code`

`params` is a polymorphic model here, thus can either have or not have `purpose` method.
This is the most compact way to write the code for a particular situation.

**Comment by MOZGIII** [comment]
This cop detected **a lot** of problems in our code, however some of them are false-positives due to this issue.

**Comment by MOZGIII** [comment]
Also, it's just perfectly fine to write something like `nil.try(...)` or `nil.try!(...)` in Rails, so if it needs to be checked it should probably be done in a separate cop.
Idea for a cop btw: use `&.` instead of `.try(...)` where possible.

**Comment by mikegee** [comment]
It would be nice if more folks contributed there thoughts on this. I'm not convinced the change requested improves the code.

In the particular case presented: `property&.params.try(:purpose)&.code`. I think defining the `purpose` method on all the objects `params` might return is a better approach.

**Comment by pocke** [comment]
it is reasonable. 


>  Idea for a cop btw: use &. instead of .try!(...) where possible.

We can use Rails/SafeNavigation cop for the purpose.
https://github.com/bbatsov/rubocop/blob/master/lib/rubocop/cop/rails/safe_navigation.rb


> In the particular case presented: property&.params.try(:purpose)&.code. I think defining the purpose method on all the objects params might return is a better approach.

Yeah, however, I think it is out of scope for this cop.
The refactoring should be suggested by other cop or other analyser.

**Comment by bbatsov** [comment]
I agree.

**Comment by igas** [comment]
Does it make sense to add more exceptions? for example:

```
foo&.bar.presence
foo&.bar.present?
foo&.bar.nil?
```

**Comment by MOZGIII** [comment]
I guess so. Any method that can be used in `nil?` is technically legal, so all of them should be allowed.

**Comment by pocke** [comment]
> Does it make sense to add more exceptions? for example:


`.present?` and `.nil?` have been added to exceptions already.  However, `.presence` hasn't been added yet.

The following methods are added by RoR into `NilClass`.


```
acts_like?
as_json
__binding__
blank?
bullet_key
byebug
class_eval
debugger
deep_dup
duplicable?
gem
html_safe?
in?
instance_values
instance_variable_names
load_dependency
presence
presence_in
present?
pretty_inspect
pretty_print
pretty_print_cycle
pretty_print_inspect
pretty_print_instance_variables
primary_key_value
pry
psych_to_yaml
require_dependency
require_or_load
suppress_warnings
to_json
to_param
to_query
to_yaml
to_yaml_properties
try
try!
unloadable
with_options
```

However, the cop supports only `blank?`, `present?` and `try`.


Should the cop support the above all methods?

**Comment by bbatsov** [comment]
Maybe we should just make this configurable and let people add there whatever they want. I myself hate a style of programming that assumes `nil`s can pop up everywhere...

**Comment by bbatsov** [comment]
At any rate - I wouldn't go overboard with this...

**Comment by MOZGIII** [comment]
I agree. It's currently almost impossible to deduce a set of methods that are allowed to be called after `&.` or `.try(...)`.
It will also vary from project to project.
But sane defaults are great anyway.

**Comment by MOZGIII** [comment]
Could you please push release to rubygems?

**Comment by bbatsov** [comment]
I don't plan to issue a new release soon. This is a tiny issue that doesn't
warrant a special release. Just build the gem from master in the mean time.

**Comment by bbatsov** [comment]
P.S. My decision might change if more bugs get fixed in the next couple of days.
